### PR TITLE
feat(ingest/bigquery): support bigquery profiling with sampling

### PR DESCRIPTION
**Pull request by mayurinehate**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/ge_profiling_config.py [hunk 1]
```diff
@@ -145,10 +145,26 @@ class GEProfilingConfig(ConfigModel):
...
+        "If not specified, profile the latest partition. Only Bigquery supports this.",
+    )
+    use_sampling: bool = Field(
+        default=True,
+        description="Whether to profile column level stats on sample of table. Only BigQuery supports this. "
+        "If enabled, profiling is done on rows sampled from table. Sampling is not done for smaller tables. ",
+    )
+
+    sample_size: int = Field(
+        default=1000,
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/ge_profiling_config.py:165]: How'd you decide on this default? Do you know approx how long each 1000, 10000, 100000, etc. rows will take? I wonder if 10,000 is a better default, for higher fidelity with a small tradeoff in speed.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 2]
```diff
@@ -961,84 +1019,18 @@ def _generate_single_profile(
...
-                raw_connection.close()
+            if custom_sql is not None:
+                # Note that limit and offset are not supported for custom SQL.
+                bq_sql = custom_sql
+            else:
+                bq_sql = f"SELECT * FROM `{table}`"
+                if self.config.limit:
+                    bq_sql += f" LIMIT {self.config.limit}"
+                if self.config.offset:
+                    bq_sql += f" OFFSET {self.config.offset}"
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:1030]: Was this just not implemented before?

**Review by asikowitz — APPROVED** [review]
LGTM! A couple questions for my own benefit and a question about the default sample size, but no blockers

**Comment by JerryDataCollins** [comment]
Hi Everyone,

 I'm not sure how you're proposing to implement data profiling by sampling but BQ (and other RDBMs) do support some functions which automatically sample the data in a table to produce sampled results. Please see this link for more details:

https://cloud.google.com/bigquery/docs/reference/standard-sql/approximate_aggregate_functions

Thanks
 Jerry

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 3]
```diff
@@ -737,6 +740,61 @@ def generate_dataset_profile(  # noqa: C901 (complexity)
...
+                self.dataset._table = sa.text(temp_table_name)
+                logger.debug(f"Setting table name to be {self.dataset._table}")
+
+                if (
+                    profile.partitionSpec
+                    and profile.partitionSpec.type == PartitionTypeClass.FULL_TABLE
+                ):
+                    profile.partitionSpec = PartitionSpecClass(
+                        type=PartitionTypeClass.QUERY, partition="SAMPLE"
+                    )
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:791]: I don't think this is represented in UI.

File: metadata-ingestion/src/datahub/ingestion/source/ge_profiling_config.py (hunk 1 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/ge_profiling_config.py:165]: I don't have performance numbers on different sizes. I'm open to changing this default. We can maybe come back to this after we have some numbers on speed and some observations on fidelity based on real world data.

**Comment by mayurinehate** [comment]
> Hi Everyone,
> 
> I'm not sure how you're proposing to implement data profiling by sampling but BQ (and other RDBMs) do support some functions which automatically sample the data in a table to produce sampled results. Please see this link for more details:
> 
> https://cloud.google.com/bigquery/docs/reference/standard-sql/approximate_aggregate_functions
> 
> Thanks Jerry

Hi Jerry, Thanks for the comment.

I've updated the PR description with implementation details. Also, Approx queries (unique count, median, quantiles) you've mentioned are always used during BigQuery Profiling, even when sampling is not enabled.

**Pull request merged**